File: c/quickSort.c
#include <stdio.h>
#include "array.h"

void quickSort(int* arr, int start, int pivot);

void main() {
    int* arr = allocateArray(); // returns an array arr[0] is always the size of arr, the actual array starts from arr[1]
    quickSort(arr + 1, 0, *arr);
    printArray(arr + 1, *arr);
}

void quickSort(int* arr, int start, int end) {
    int pivot = end - 1, pos = start, i;
    for(i = start; i < pivot; i++) {
        if(arr[i] < arr[pivot]) {
            pos++;
        }
    }
    if(pos != pivot) {
        arr[pos] += arr[pivot];
        arr[pivot] = arr[pos] - arr[pivot];
        arr[pos] -= arr[pivot];
    }
    pivot = pos;
    i = start;
    pos = pivot + 1;
    while(i < pivot) {
        if(arr[i] > arr[pivot] && arr[pos] <= arr[pivot]) {
            arr[i] += arr[pos];
            arr[pos] = arr[i] - arr[pos];
            arr[i] -= arr[pos];
        }
        else if (arr[i] > arr[pivot] && arr[pos] > arr[pivot]) {
            pos++;
            i--;
        }
        i++;
    }
    if((pivot - start) > 1) {
        quickSort(arr, start, pivot);
    }
    if((end - pivot) > 1){
        quickSort(arr, pivot + 1, end);
    }
}

File: python/quickSort.py
def quickSort(start, end, li):
    if start < end:
        pivot = start
        for i in range(start, end):
            if li[i] < li[(start + end) // 2]:
                pivot += 1
        if pivot != (start + end) // 2:
            li[(start + end) // 2] += li[pivot]
            li[pivot] = li[(start + end) // 2] - li[pivot]
            li[(start + end) // 2] -= li[pivot]
        i = start
        j = end - 1
        while i < j:
            if li[i] >= li[pivot] and li[pivot] > li[j]:
                li[i] += li[j]
                li[j] = li[i] - li[j]
                li[i] -= li[j]
                i += 1
            elif li[i] > li[pivot] and li[pivot] < li[j] or pivot == j:
                j -= 1
            else:
                i += 1
        quickSort(start, pivot, li)
        quickSort(pivot + 1, end, li)



li = [6,5,4,3,2,1]
# li = [7,8,5,6,4,3]
# li = [6,2,1,4,5,7,8]
# li = [7,2,1,6,8,5,3,4]
# li = [1,1,1,1,1,1]
# li = [1,2,3,3,2,1]

quickSort(0, len(li), li)
print(li)
